File: nhglib-core/src/io/github/movementspeed/nhglib/core/ecs/components/scenes/NodeComponent.java
package io.github.movementspeed.nhglib.core.ecs.components.scenes;

import com.artemis.PooledComponent;
import com.badlogic.gdx.graphics.g3d.model.Node;
import com.badlogic.gdx.math.Matrix4;
import com.badlogic.gdx.math.Quaternion;
import com.badlogic.gdx.math.Vector3;

/**
 * Created by Fausto Napoli on 08/12/2016.
 * The base component needed by the SceneGraph for entity positioning.
 * Should only be created by the SceneGraph together with an entity.
 */
public class NodeComponent extends PooledComponent {
    /**
     * Unique NodeComponent ID, corresponds with entity itself
     */
    public int id;

    /**
     * Node
     */
    public Node node;

    /**
     * Parent NodeComponent, can be null.
     */
    public NodeComponent parentNodeComponent;

    public String parentInternalNodeId;

    private Vector3 tempVec;
    private Vector3 tempVec2;
    private Quaternion tempQuat;

    private Vector3 translation;
    private Vector3 rotation;
    private Vector3 scale;

    private Vector3 translationDelta;
    private Vector3 rotationDelta;
    private Vector3 scaleDelta;

    private Quaternion rotationQuaternion;

    public NodeComponent() {
        node = new Node();

        tempVec = new Vector3();
        tempVec2 = new Vector3();
        tempQuat = new Quaternion();

        translation = new Vector3();
        rotation = new Vector3();
        scale = new Vector3(1, 1, 1);

        translationDelta = new Vector3();
        rotationDelta = new Vector3();
        scaleDelta = new Vector3();

        rotationQuaternion = new Quaternion();
    }

    @Override
    protected void reset() {
        node.translation.set(new Vector3());
        node.rotation.set(new Quaternion());
        node.scale.set(new Vector3());

        translationDelta.set(Vector3.Zero);
        rotationDelta.set(Vector3.Zero);
        scaleDelta.set(Vector3.Zero);
    }

    public void setId(int id) {
        this.id = id;
        node.id = "node_" + id;
    }

    public void setTranslation(Vector3 translation) {
        setTranslation(translation, false);
    }

    public void setTranslation(Vector3 translation, boolean apply) {
        setTranslation(translation.x, translation.y, translation.z, apply);
    }

    public void setTranslation(float x, float y, float z) {
        setTranslation(x, y, z, false);
    }

    public void setTranslation(float x, float y, float z, boolean apply) {
        translationDelta.set(
                translation.x - x,
                translation.y - y,
                translation.z - z);

        node.translation.set(x, y, z);

        if (apply) {
            applyTransforms();
        }
    }

    public void setTranslationX(float x) {
        translationDelta.x = translation.x - x;
        node.translation.x = x;
    }

    public void setTranslationY(float y) {
        translationDelta.y = translation.y - y;
        node.translation.y = y;
    }

    public void setTranslationZ(float z) {
        translationDelta.z = translation.z - z;
        node.translation.z = z;
    }

    public void translate(Vector3 translation) {
        translate(translation, false);
    }

    public void translate(Vector3 translation, boolean apply) {
        translate(translation.x, translation.y, translation.z, apply);
    }

    public void translate(float x, float y, float z) {
        translate(x, y, z, false);
    }

    /**
     * Translates the node.
     *
     * @param x     translation.
     * @param y     translation.
     * @param z     translation.
     * @param apply if true, transforms will be calculated immediately. It's not recommended, instead use
     *              {@link #applyTransforms() applyTransforms()} after you've completed all transforms on the node.
     */
    public void translate(float x, float y, float z, boolean apply) {
        translationDelta.set(
                translation.x - x,
                translation.y - y,
                translation.z - z);

        tempVec.set(x, y, z);

        float len = tempVec.len();
        tempVec.rot(node.localTransform).nor().scl(len);

        node.translation.add(tempVec);

        if (apply) {
            applyTransforms();
        }
    }

    public void setRotation(Vector3 rotation) {
        setRotation(rotation, false);
    }

    public void setRotation(Vector3 rotation, boolean apply) {
        setRotation(rotation.x, rotation.y, rotation.z, apply);
    }

    public void setRotation(float x, float y, float z) {
        setRotation(x, y, z, false);
    }

    public void setRotation(float x, float y, float z, boolean apply) {
        rotationDelta.set(
                rotation.x - x,
                rotation.y - y,
                rotation.z - z);

        node.rotation.setEulerAngles(y, x, z);

        if (apply) {
            applyTransforms();
        }
    }

    public void setRotation(Quaternion rotation) {
        rotationDelta.set(
                rotation.x - rotation.getPitch(),
                rotation.y - rotation.getYaw(),
                rotation.z - rotation.getRoll());

        node.rotation.set(rotation);
    }

    public void rotate(Quaternion quaternion) {
        rotate(quaternion, false);
    }

    /**
     * Rotates the node.
     *
     * @param rotation the rotation quaternion added to the node's rotation.
     * @param apply    if true, transforms will be calculated immediately. It's not recommended, instead use
     *                 {@link #applyTransforms() applyTransforms()} after you've completed all transforms on the node.
     */
    public void rotate(Quaternion rotation, boolean apply) {
        node.rotation.add(rotation);

        if (apply) {
            applyTransforms();
        }
    }

    public void rotate(Vector3 rotation) {
        rotate(rotation, false);
    }

    public void rotate(Vector3 rotation, boolean apply) {
        rotate(rotation.x, rotation.y, rotation.z, apply);
    }

    public void rotate(float x, float y, float z) {
        rotate(x, y, z, false);
    }

    /**
     * Rotates the node.
     *
     * @param x     rotation, also known as pitch.
     * @param y     rotation, also known as yaw.
     * @param z     rotation, also known as roll.
     * @param apply if true, transforms will be calculated immediately. It's not recommended, instead use
     *              {@link #applyTransforms() applyTransforms()} after you've completed all transforms on the node.
     */
    public void rotate(float x, float y, float z, boolean apply) {
        rotationDelta.set(
                rotation.x - x,
                rotation.y - y,
                rotation.z - z);

        tempQuat.setEulerAngles(y, x, z);
        node.rotation.mul(tempQuat);

        if (apply) {
            applyTransforms();
        }
    }

    public void setScale(Vector3 scale) {
        setScale(scale, false);
    }

    public void setScale(Vector3 scale, boolean apply) {
        setScale(scale.x, scale.y, scale.z, apply);
    }

    public void setScale(float x, float y, float z) {
        setScale(x, y, z, false);
    }

    public void setScale(float x, float y, float z, boolean apply) {
        scaleDelta.set(
                scale.x - x,
                scale.y - y,
                scale.z - z);

        node.scale.set(x, y, z);

        if (apply) {
            applyTransforms();
        }
    }

    public void scale(Vector3 scale) {
        scale(scale, false);
    }

    public void scale(Vector3 scale, boolean apply) {
        scale(scale.x, scale.y, scale.y, apply);
    }

    public void scale(float x, float y, float z) {
        scale(x, y, z, false);
    }

    /**
     * Scales the node.
     *
     * @param x     scale.
     * @param y     scale.
     * @param z     scale.
     * @param apply if true, transforms will be calculated immediately. It's not recommended, instead use
     *              {@link #applyTransforms() applyTransforms()} after you've completed all transforms on the node.
     */
    public void scale(float x, float y, float z, boolean apply) {
        scaleDelta.set(
                scale.x - x,
                scale.y - y,
                scale.z - z);

        node.scale.add(x, y, z);

        if (apply) {
            applyTransforms();
        }
    }

    public void setTransform(Matrix4 transform) {
        if (transform != null) {
            transform.getTranslation(tempVec);
            transform.getRotation(tempQuat);
            transform.getScale(tempVec2);

            setTranslation(tempVec);
            setRotation(tempQuat);
            setScale(tempVec2);
        }
    }

    public void setTransform(Vector3 translation, Vector3 rotation, Vector3 scale) {
        setTranslation(translation);
        setRotation(rotation);
        setScale(scale);
    }

    public void applyTransforms() {
        applyTransforms(true);
    }

    public void applyTransforms(boolean recursive) {
        node.calculateTransforms(recursive);
    }

    public float getX() {
        return getTranslation().x;
    }

    public float getY() {
        return getTranslation().y;
    }

    public float getZ() {
        return getTranslation().z;
    }

    public float getXRotation() {
        return getRotation().x;
    }

    public float getYRotation() {
        return getRotation().y;
    }

    public float getZRotation() {
        return getRotation().z;
    }

    public float getXScale() {
        return getScale().x;
    }

    public float getYScale() {
        return getScale().y;
    }

    public float getZScale() {
        return getScale().z;
    }

    public float getXDelta() {
        float res = translationDelta.x;
        translationDelta.x = 0;
        return res;
    }

    public float getYDelta() {
        float res = translationDelta.y;
        translationDelta.y = 0;
        return res;
    }

    public float getZDelta() {
        float res = translationDelta.z;
        translationDelta.z = 0;
        return res;
    }

    public float getXRotationDelta() {
        float res = rotationDelta.x;
        rotationDelta.x = 0;
        return res;
    }

    public float getYRotationDelta() {
        float res = rotationDelta.y;
        rotationDelta.y = 0;
        return res;
    }

    public float getZRotationDelta() {
        float res = rotationDelta.z;
        rotationDelta.z = 0;
        return res;
    }

    public float getXScaleDelta() {
        float res = scaleDelta.x;
        scaleDelta.x = 0;
        return res;
    }

    public float getYScaleDelta() {
        float res = scaleDelta.y;
        scaleDelta.y = 0;
        return res;
    }

    public float getZScaleDelta() {
        float res = scaleDelta.z;
        scaleDelta.z = 0;
        return res;
    }

    public Matrix4 getTransform() {
        return node.globalTransform;
    }

    public Vector3 getTranslation() {
        node.globalTransform.getTranslation(translation);
        return translation;
    }

    public Vector3 getRotation() {
        node.globalTransform.getRotation(rotationQuaternion);
        rotation.set(
                rotationQuaternion.getPitch(),
                rotationQuaternion.getYaw(),
                rotationQuaternion.getRoll());

        return rotation;
    }

    public Vector3 getScale() {
        node.globalTransform.getScale(scale);
        return scale;
    }

    public Vector3 getTranslationDelta() {
        Vector3 res = tempVec.set(translationDelta);
        translationDelta.set(Vector3.Zero);

        return res;
    }

    public Vector3 getRotationDelta() {
        Vector3 res = tempVec.set(rotationDelta);
        rotationDelta.set(Vector3.Zero);

        return res;
    }

    public Vector3 getScaleDelta() {
        Vector3 res = tempVec.set(scaleDelta);
        scaleDelta.set(Vector3.Zero);

        return res;
    }

    public Quaternion getRotationQuaternion() {
        getRotation();
        return rotationQuaternion;
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/core/ecs/components/scenes/NodeComponent.kt
package io.github.movementspeed.nhglib.core.ecs.components.scenes

import com.artemis.PooledComponent
import com.badlogic.gdx.graphics.g3d.model.Node
import com.badlogic.gdx.math.Matrix4
import com.badlogic.gdx.math.Quaternion
import com.badlogic.gdx.math.Vector3

/**
 * Created by Fausto Napoli on 08/12/2016.
 * The base component needed by the SceneGraph for entity positioning.
 * Should only be created by the SceneGraph together with an entity.
 */
class NodeComponent : PooledComponent() {
    /**
     * Unique NodeComponent ID, corresponds with entity itself
     */
    var id: Int = 0
        set(value) {
            field = value
            node.id = "node_$value"
        }

    /**
     * Node
     */
    var node = Node()

    /**
     * Parent NodeComponent, can be null.
     */
    var parentNodeComponent: NodeComponent? = null

    var parentInternalNodeId: String? = null

    private val tempVec = Vector3()
    private val tempVec2 = Vector3()
    private val tempQuat = Quaternion()

    private val translation = Vector3()
    private val rotation = Vector3()
    private val scale = Vector3(1f, 1f, 1f)

    private val localTranslation = Vector3()
    private val localRotation = Vector3()
    private val localScale = Vector3(1f, 1f, 1f)

    private val translationDelta = Vector3()
    private val rotationDelta = Vector3()
    private val scaleDelta = Vector3()

    private val localRotationQuaternion = Quaternion()
    private val rotationQuaternion = Quaternion()

    val localX: Float
        get() = getLocalTranslation().x

    val localY: Float
        get() = getLocalTranslation().y

    val localZ: Float
        get() = getLocalTranslation().z

    val localXRotation: Float
        get() = getLocalRotation().x

    val localYRotation: Float
        get() = getLocalRotation().y

    val localZRotation: Float
        get() = getLocalRotation().z

    val localXScale: Float
        get() = getLocalScale().x

    val localYScale: Float
        get() = getLocalScale().y

    val localZScale: Float
        get() = getLocalScale().z

    val x: Float
        get() = getTranslation().x

    val y: Float
        get() = getTranslation().y

    val z: Float
        get() = getTranslation().z

    val xRotation: Float
        get() = getRotation().x

    val yRotation: Float
        get() = getRotation().y

    val zRotation: Float
        get() = getRotation().z

    val xScale: Float
        get() = getScale().x

    val yScale: Float
        get() = getScale().y

    val zScale: Float
        get() = getScale().z

    val xDelta: Float
        get() {
            val res = translationDelta.x
            translationDelta.x = 0f
            return res
        }

    val yDelta: Float
        get() {
            val res = translationDelta.y
            translationDelta.y = 0f
            return res
        }

    val zDelta: Float
        get() {
            val res = translationDelta.z
            translationDelta.z = 0f
            return res
        }

    val xRotationDelta: Float
        get() {
            val res = rotationDelta.x
            rotationDelta.x = 0f
            return res
        }

    val yRotationDelta: Float
        get() {
            val res = rotationDelta.y
            rotationDelta.y = 0f
            return res
        }

    val zRotationDelta: Float
        get() {
            val res = rotationDelta.z
            rotationDelta.z = 0f
            return res
        }

    val xScaleDelta: Float
        get() {
            val res = scaleDelta.x
            scaleDelta.x = 0f
            return res
        }

    val yScaleDelta: Float
        get() {
            val res = scaleDelta.y
            scaleDelta.y = 0f
            return res
        }

    val zScaleDelta: Float
        get() {
            val res = scaleDelta.z
            scaleDelta.z = 0f
            return res
        }

    val localTransform: Matrix4
        get() = node.localTransform

    var transform: Matrix4?
        get() = node.globalTransform
        set(transform) {
            if (transform != null) {
                transform.getTranslation(tempVec)
                transform.getRotation(tempQuat)
                transform.getScale(tempVec2)

                setTranslation(tempVec)
                setRotation(tempQuat)
                setScale(tempVec2)
            }
        }

    override fun reset() {
        node.translation.set(Vector3())
        node.rotation.set(Quaternion())
        node.scale.set(Vector3())

        translationDelta.set(Vector3.Zero)
        rotationDelta.set(Vector3.Zero)
        scaleDelta.set(Vector3.Zero)
    }

    @JvmOverloads
    fun setTranslation(translation: Vector3, apply: Boolean = false) {
        setTranslation(translation.x, translation.y, translation.z, apply)
    }

    @JvmOverloads
    fun setTranslation(x: Float, y: Float, z: Float, apply: Boolean = false) {
        translationDelta.set(
                translation.x - x,
                translation.y - y,
                translation.z - z)

        node.translation.set(x, y, z)

        if (apply) {
            applyTransforms()
        }
    }

    fun setTranslationX(x: Float) {
        translationDelta.x = translation.x - x
        node.translation.x = x
    }

    fun setTranslationY(y: Float) {
        translationDelta.y = translation.y - y
        node.translation.y = y
    }

    fun setTranslationZ(z: Float) {
        translationDelta.z = translation.z - z
        node.translation.z = z
    }

    @JvmOverloads
    fun translate(translation: Vector3, apply: Boolean = false) {
        translate(translation.x, translation.y, translation.z, apply)
    }

    /**
     * Translates the node.
     *
     * @param x     translation.
     * @param y     translation.
     * @param z     translation.
     * @param apply if true, transforms will be calculated immediately. It's not recommended, instead use
     * [applyTransforms()][.applyTransforms] after you've completed all transforms on the node.
     */
    @JvmOverloads
    fun translate(x: Float, y: Float, z: Float, apply: Boolean = false) {
        translationDelta.set(
                translation.x - x,
                translation.y - y,
                translation.z - z)

        tempVec.set(x, y, z)

        val len = tempVec.len()
        tempVec.rot(node.localTransform).nor().scl(len)

        node.translation.add(tempVec)

        if (apply) {
            applyTransforms()
        }
    }

    @JvmOverloads
    fun setRotation(rotation: Vector3, apply: Boolean = false) {
        setRotation(rotation.x, rotation.y, rotation.z, apply)
    }

    @JvmOverloads
    fun setRotation(x: Float, y: Float, z: Float, apply: Boolean = false) {
        rotationDelta.set(
                rotation.x - x,
                rotation.y - y,
                rotation.z - z)

        node.rotation.setEulerAngles(y, x, z)

        if (apply) {
            applyTransforms()
        }
    }

    fun setRotation(rotation: Quaternion) {
        rotationDelta.set(
                rotation.x - rotation.pitch,
                rotation.y - rotation.yaw,
                rotation.z - rotation.roll)

        node.rotation.set(rotation)
    }

    fun rotate(quaternion: Quaternion) {
        rotate(quaternion, false)
    }

    /**
     * Rotates the node.
     *
     * @param rotation the rotation quaternion added to the node's rotation.
     * @param apply    if true, transforms will be calculated immediately. It's not recommended, instead use
     * [applyTransforms()][.applyTransforms] after you've completed all transforms on the node.
     */
    fun rotate(rotation: Quaternion, apply: Boolean) {
        node.rotation.add(rotation)

        if (apply) {
            applyTransforms()
        }
    }

    @JvmOverloads
    fun rotate(rotation: Vector3, apply: Boolean = false) {
        rotate(rotation.x, rotation.y, rotation.z, apply)
    }

    /**
     * Rotates the node.
     *
     * @param x     rotation, also known as pitch.
     * @param y     rotation, also known as yaw.
     * @param z     rotation, also known as roll.
     * @param apply if true, transforms will be calculated immediately. It's not recommended, instead use
     * [applyTransforms()][.applyTransforms] after you've completed all transforms on the node.
     */
    @JvmOverloads
    fun rotate(x: Float, y: Float, z: Float, apply: Boolean = false) {
        rotationDelta.set(
                rotation.x - x,
                rotation.y - y,
                rotation.z - z)

        tempQuat.setEulerAngles(y, x, z)
        node.rotation.mul(tempQuat)

        if (apply) {
            applyTransforms()
        }
    }

    @JvmOverloads
    fun setScale(scale: Vector3, apply: Boolean = false) {
        setScale(scale.x, scale.y, scale.z, apply)
    }

    @JvmOverloads
    fun setScale(x: Float, y: Float, z: Float, apply: Boolean = false) {
        scaleDelta.set(
                scale.x - x,
                scale.y - y,
                scale.z - z)

        node.scale.set(x, y, z)

        if (apply) {
            applyTransforms()
        }
    }

    @JvmOverloads
    fun scale(scale: Vector3, apply: Boolean = false) {
        scale(scale.x, scale.y, scale.y, apply)
    }

    /**
     * Scales the node.
     *
     * @param x     scale.
     * @param y     scale.
     * @param z     scale.
     * @param apply if true, transforms will be calculated immediately. It's not recommended, instead use
     * [applyTransforms()][.applyTransforms] after you've completed all transforms on the node.
     */
    @JvmOverloads
    fun scale(x: Float, y: Float, z: Float, apply: Boolean = false) {
        scaleDelta.set(
                scale.x - x,
                scale.y - y,
                scale.z - z)

        node.scale.add(x, y, z)

        if (apply) {
            applyTransforms()
        }
    }

    fun setTransform(translation: Vector3, rotation: Vector3, scale: Vector3) {
        setTranslation(translation)
        setRotation(rotation)
        setScale(scale)
    }

    @JvmOverloads
    fun applyTransforms(recursive: Boolean = true) {
        node.calculateTransforms(recursive)
    }

    fun getLocalTranslation(): Vector3 {
        node.localTransform.getTranslation(localTranslation)
        return localTranslation
    }

    fun getTranslation(): Vector3 {
        node.globalTransform.getTranslation(translation)
        return translation
    }

    fun getLocalRotation(): Vector3 {
        node.localTransform.getRotation(localRotationQuaternion)
        localRotation.set(
                localRotationQuaternion.pitch,
                localRotationQuaternion.yaw,
                localRotationQuaternion.roll)

        return localRotation
    }

    fun getRotation(): Vector3 {
        node.globalTransform.getRotation(rotationQuaternion)
        rotation.set(
                rotationQuaternion.pitch,
                rotationQuaternion.yaw,
                rotationQuaternion.roll)

        return rotation
    }

    fun getLocalScale(): Vector3 {
        node.localTransform.getScale(localScale)
        return localScale
    }

    fun getScale(): Vector3 {
        node.globalTransform.getScale(scale)
        return scale
    }

    fun getTranslationDelta(): Vector3 {
        val res = tempVec.set(translationDelta)
        translationDelta.set(Vector3.Zero)

        return res
    }

    fun getRotationDelta(): Vector3 {
        val res = tempVec.set(rotationDelta)
        rotationDelta.set(Vector3.Zero)

        return res
    }

    fun getScaleDelta(): Vector3 {
        val res = tempVec.set(scaleDelta)
        scaleDelta.set(Vector3.Zero)

        return res
    }

    fun getLocalRotationQuaternion(): Quaternion {
        getLocalRotation()
        return localRotationQuaternion
    }

    fun getRotationQuaternion(): Quaternion {
        getRotation()
        return rotationQuaternion
    }
}
